docs/algorithm/algorithm.tex

```diff
@@ -0,0 +1,383 @@
+of the frequency changes for the continuous clock change. However, if neither is needed after doing a measurement, a seperate correction for the frequency
+is done. With the standard settings, this corrects away the entire observed frequency error.
+
+In ntpd-rs, all the numbers named above can be configured to have different values, and for frequency a similar threshold-leftover scheme can be used as for offset.
+The defaults for the latter are just both set to 0.
+
+\section{Poll interval selection}
+
+There is one final thing we haven't touched on yet, and that is how ntpd-rs chooses how often to ask each server for the time. This is done in two stages. First, the algorithm selects a desired poll interval for all time sources. Then, on a per time source basis, this is adapted to deal with server-indicated limits on how often we may ask, and backoffs that kick in on failures. Here, we will focus only on the selection of a desired poll interval.
+
+For this, first a desired poll interval is selected for each time source such that the measurement noise is between 40 and 60 percent of the total uncertainty on the difference between prediction and measurement. This is again done with a counter mechanism similar to the process noise estimation. This time, any measurement for which less than 40 percent of the difference uncertainty is measurement is a vote for a longer poll interval, anything where it is more than 60 percent is a vote for a shorter poll interval and anything in between pushes the counter back to 0.
+
+Then the overal desired poll interval is the minimum of the desired poll intervals of all the time sources. This is currently done as a global indicator primarily because of limitations in the architecture of ntpd-rs. In the future, this may change to a desired poll interval that is different for each peer.
+
+Note also that, as in the case of process noise estimation, the specific numbers mentioned above are configurable. The values above are the current default values used by ntpd-rs.
+
+\section{Performance}
+
+We evaluated the performance of the new algorithm against the default settings of chrony version 4.0, another NTP client. For this, we connected a raspberry pi 4 via a switch to a local gps-connected ntp server (an Endrun Ninja). Both the raspberry pi and the ntp server were configured to produce pulse per second outputs.
+
+For each test, the NTP client under test was started on the raspberry pi with a fixed poll interval of 1 second and allowed to stabilize over a period of at least half an hour. Then, the offset between the pulse per second signals of the raspberry pi were measured for one hour. The results of these measurements are shown in Figure~\ref{fig:offset-time}. Both chrony and ntpd-rs were configured to use the local server as only timesource, but otherwise used their default configuration.
```

Comment:
```suggestion
For each test, the NTP client under test was started on the Raspberry Pi with a fixed poll interval of 1 second and allowed to stabilize over a period of at least half an hour. Then, the offset between the pulse per second signals of the Raspberry Pi were measured for one hour. The results of these measurements are shown in Figure~\ref{fig:offset-time}. Both chrony and ntpd-rs were configured to use the local server as only timesource, but otherwise used their default configuration.
```